src/hotspot/share/opto/type.hpp

```diff
@@ -364,6 +364,17 @@ class Type {
   }
   virtual void dump2( Dict &d, uint depth, outputStream *st ) const;
   static  void dump_stats();
+  // Groups of types, for debugging and visualization only.
+  enum CATEGORY {
+    CatData,
+    CatMemory,
+    CatMixed,   // Tuples with types of different categories.
+    CatControl,
+    CatOther,   // {Type::Top, Type::Abio, Type::Bottom}.
+    CatUndef    // {Type::Bad, Type::lastype}, for completeness.
```

Comment:
Is "Cat" required when the enum is already called "CATEGORY"?

Reply:
Right, it was required in a C-style unscoped-enum because of collisions with other enums in type.hpp. But I found in the HotSpot Style Guide that scoped-enums are actually encouraged, so I changed to that.